src/dune_pkg/project_source_stanza.ml

```diff
@@ -0,0 +1,295 @@
+        ; Pp.textf "it's already assigned to %s at:" (OpamUrl.to_string (snd package.url))
+        ; Pp.verbatim (Loc.to_file_colon_line assigned.loc)
+        ]
+  in
+  let opam_package (package : Local_package.pin) =
+    let* resolved_package = Pinned_package.resolve_package package in
+    resolve package.name resolved_package;
+    Resolved_package.opam_file resolved_package
+    |> OpamFile.OPAM.pin_depends
+    |> Fiber.parallel_iter ~f:(fun (pkg, url) ->
+      let version = OpamPackage.version pkg |> Package_version.of_opam_package_version in
+      let name = Package_name.of_opam_package_name (OpamPackage.name pkg) in
+      let package =
+        { Local_package.url = package.loc, url
+        ; version
+        ; name
+        ; loc = package.loc
+        ; kind = `Opam
+        }
+      in
+      assign package (fun () -> Pinned_package.resolve_package package >>| resolve name))
+  in
+  let dune_package packages (package : Local_package.pin) =
+    match Package_name.Map.find packages package.name with
+    | None ->
+      User_error.raise
+        ~loc:package.loc
+        [ Pp.textf
+            "package %s doesn't exist in source %s"
+            (Package_name.to_string package.name)
+            (OpamUrl.to_string (snd package.url))
+        ]
+    | Some pkg ->
+      let opam_file =
+        Local_package.for_solver pkg
+        |> Local_package.For_solver.to_opam_file
+        |> OpamFile.OPAM.with_url (OpamFile.URL.create (snd package.url))
+        |> OpamFile.OPAM.with_build
+             [ (* CR-rgrinberg: this needs to be addressed in a more principled manner *)
+               (let cmd =
+                  ([ "dune"; "build"; "-p" ] |> List.map ~f:(fun x -> OpamTypes.CString x))
+                  @ [ OpamTypes.CIdent "name" ]
+                  |> List.map ~f:(fun x -> x, None)
+                in
+                cmd, None)
+             ]
+      in
+      let opam_package =
+        OpamPackage.create
+          (Package_name.to_opam_package_name package.name)
+          (Package_version.to_opam_package_version package.version)
+      in
+      let resolved_package =
+        Resolved_package.dune_package package.loc opam_file opam_package
+      in
+      resolve package.name resolved_package
+  in
+  let rec loop (stack : Local_package.pin list) t =
+    Package_name.Map.values t.map
+    |> Fiber.parallel_iter ~f:(fun (package, _) ->
+      let stack = Stack.push stack package in
+      assign package
+      @@ fun () ->
+      match package.kind with
+      | `Opam -> opam_package package
+      | `Dune ->
+        Scan_project.eval_url scan_project package.url
+        >>= (function
+         | None -> opam_package package
+         | Some (more_sources, packages) ->
+           dune_package packages package;
+           let more_sources = DB.add_opam_pins more_sources packages in
+           loop stack more_sources))
+  in
+  let+ () = loop [] t in
+  !resolved
```

Comment:
It's just the simplest. Otherwise we'd need to return the assign packages at every step and then merge them somehow. And the end result would detect errors less promptly without some special data structures.

If you want to try and avoid mutation, feel to try and refactor it later.